pulsar4s-akka-streams/src/main/scala/com/sksamuel/pulsar4s/akka/streams/PulsarSourceGraphStage.scala

```diff
@@ -92,24 +94,30 @@ class PulsarSourceGraphStage[T](create: () => Consumer[T], seek: Option[MessageI
       }
 
       override def onPull(): Unit = {
-        implicit val context: ExecutionContext = super.materializer.executionContext
+        implicit val context: ExecutionContext = materializer.executionContext
         logger.debug("Pull received; asking consumer for message")
         consumer.receiveAsync.onComplete {
           case Success(msg) =>
             logger.debug(s"Msg received $msg")
-            callback.invoke(msg)
+            receiveCallback.invoke(msg)
             consumer.acknowledge(msg.messageId)
           case Failure(e) =>
             logger.warn("Error when receiving message", e)
             failStage(e)
         }
       }
 
-      override def stop(): Unit = completeStage()
+      private def close()(implicit ec: ExecutionContext): Future[Done] =
+        consumerOpt.fold(Future.successful(Done))(_.closeAsync.map(_ => Done))
+
+      // Note: unlike the committable source, we don't expect acks so we can close immediately
+      override def postStop(): Unit = close()(materializer.executionContext)
```

Comment:
So previously we were never shutting down the pulsar consumer unless the user did it explicitly via shutdown?

Reply:
That's right. If you only invoked `stop` or the stage was failed or canceled, the consumer wouldn't get closed.